### PR TITLE
CA1010: Allows users to provide additional interfaces

### LINKED ISSUE #1490: CA1010 should analyze IDictionary as well
Currently, CA1010 only reports on IList, ICollection, and IEnumerable implementations missing the generic version. We should also look into reporting on dictionary types as well. The original FxCop implementation disregards dictionary types "to reduce noise", http://index/?query=CollectionsShouldImplemen&rightProject=DesignRules&file=CollectionsShouldImplementGenericInterface.cs&line=24, but we should investigate if this will actually make the analyzer more noisy.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #1490

### CODE REVIEW

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs [hunk 1]
```diff
@@ -128,57 +128,57 @@ static bool TryParseValue(string value, out ImmutableHashSet<TEnum> result)
             return analyzerConfigOptions.GetOptionValue(optionName, rule, uint.TryParse, defaultValue);
         }
 
-        public static SymbolNamesOption GetNullCheckValidationMethodsOption(
+        public static SymbolNamesOption<byte> GetNullCheckValidationMethodsOption(
```
Comment (Evangelink) [inline src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: I am not a big fan of the `<byte>` but I couldn't find anything nice except for adding a `Unit` class.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs [hunk 2]
```diff
@@ -95,95 +109,41 @@ public override void Initialize(AnalysisContext analysisContext)
                 return;
             }
 
-            var allInterfacesStatus = default(CollectionsInterfaceStatus);
-            foreach (var @interface in namedTypeSymbol.AllInterfaces)
+            var allInterfaces = ImmutableHashSet.CreateRange(namedTypeSymbol.AllInterfaces.Select(i => i.OriginalDefinition));
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]:
```suggestion
            using var allInterfaces = PooledHashSet.GetInstance(namedTypeSymbol.AllInterfaces.Select(i => i.OriginalDefinition));
```

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs [hunk 3]
```diff
@@ -193,41 +193,80 @@ static SymbolNamesOption.NameParts GetParts(string name)
...
+                }
+
+                var matchingSymbols = DocumentationCommentId.GetSymbolsForDeclarationId(genericInterfaceFullName, compilation);
+
+                if (matchingSymbols.Length != 1 ||
+                    !(matchingSymbols[0] is INamedTypeSymbol namedType) ||
+                    namedType.TypeKind != TypeKind.Interface ||
+                    !namedType.IsGenericType)
+                {
+                    // Invalid matching type so we assume there was no associated type
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: We may want to eventually report a diagnostic here so user can fix his configuration. It is fine to not do it here as it seems applicable for all our option values.

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs [hunk 4]
```diff
@@ -128,57 +128,57 @@ static bool TryParseValue(string value, out ImmutableHashSet<TEnum> result)
...
+        public static SymbolNamesOption<byte> GetExcludedTypeNamesWithDerivedTypesOption(
             this AnalyzerOptions options,
             DiagnosticDescriptor rule,
             Compilation compilation,
             CancellationToken cancellationToken)
-            => options.GetSymbolNamesOption(EditorConfigOptionNames.ExcludedTypeNamesWithDerivedTypes, namePrefixOpt: "T:", rule, compilation, cancellationToken);
+            => options.GetSymbolNamesOption<byte>(EditorConfigOptionNames.ExcludedTypeNamesWithDerivedTypes, namePrefixOpt: "T:", rule, compilation, cancellationToken);
 
-        public static SymbolNamesOption GetDisallowedSymbolNamesOption(
+        public static SymbolNamesOption<byte> GetDisallowedSymbolNamesOption(
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: I am wondering if the base type can be a non-generic type `SymbolNamesOption` and the sub-type can be `SymbolNamesOption<T>` so only the options which needs a key-value mapping will have to specify the type argument. What do you think?

Comment (mavasani) [reply src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: Probably also call it `SymbolNamesWithValuesOption<T>` or some such and move the `TryGetValue` function to this type.

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs (hunk 3 above)
Comment (Evangelink) [reply src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: #3267

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs (hunk 2 above)
Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]: Suggested overload doesn't exist but I will add it.

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs (hunk 4 above)
Comment (Evangelink) [reply src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: The rename is a good point. Regarding the base class not generic, I was thinking about that but I couldn't managed to do some implementation without too much duplication. Actually I did, but that was making the base handling a dictionary with `object` as value and so potentially forcing some unneeded boxing. Feel free to make a suggestion if you think I am missing something.

File: src/Microsoft.CodeAnalysis.AnalyzerUtilities/PublicAPI.Unshipped.txt [hunk 5]
```diff
@@ -709,3 +709,12 @@ virtual Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DataFlowOperationVisitor<TA
 virtual Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DataFlowOperationVisitor<TAnalysisData, TAnalysisContext, TAnalysisResult, TAbstractAnalysisValue>.VisitUnaryOperatorCore(Microsoft.CodeAnalysis.Operations.IUnaryOperation operation, object argument) -> TAbstractAnalysisValue
 virtual Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PredicatedAnalysisData<TKey, TValue>.ApplyPredicatedData(Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DictionaryAnalysisData<TKey, TValue> coreAnalysisData, Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DictionaryAnalysisData<TKey, TValue> predicatedData) -> void
 virtual Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PredicatedAnalysisData<TKey, TValue>.RemoveEntryInPredicatedData(TKey key, Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DictionaryAnalysisData<TKey, TValue> predicatedData) -> void
+Analyzer.Utilities.Unit
```
Comment (mavasani) [inline src/Microsoft.CodeAnalysis.AnalyzerUtilities/PublicAPI.Unshipped.txt]: Please make this an internal type and revert the changes to the PublicAPI files. We only want to expose public APIs for functionality, not for utilities from this package.

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs (hunk 4 above)
Comment (mavasani) [reply src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: Can we have an `AbstractSymbolNamesOption` which has all the core logic, with sub-types `SymbolNamesOption` and `SymbolNamesWithValueOption<T>` maintaining their required datastructures (set in the former, and dictionary for the latter)?

Comment (Evangelink) [reply src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: We can but there won't be much code common between the 2, I'll add a separate commit so you can decide what you think is best.

File: src/Microsoft.CodeAnalysis.AnalyzerUtilities/PublicAPI.Unshipped.txt (hunk 5 above)
Comment (Evangelink) [reply src/Microsoft.CodeAnalysis.AnalyzerUtilities/PublicAPI.Unshipped.txt]: Makes sense, I didn't pay attention here...

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs (hunk 4 above)
Comment (mavasani) [reply src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: Ah, in that case it is fine to keep it as is.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs [hunk 6]
```diff
@@ -95,95 +110,45 @@ public override void Initialize(AnalysisContext analysisContext)
...
+
+                if (allInterfaces.Contains(kvp.Key) && !allInterfaces.Contains(kvp.Value))
+                {
+                    ReportDiagnostic(kvp.Key, kvp.Value);
+                    return;
+                }
             }
-            else
+
+            void ReportDiagnostic(INamedTypeSymbol @interface, INamedTypeSymbol genericInterface)
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]: Nit: please add an explicit return statement before start of local function definitions.

Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]: Is it for reading purpose or is there any impact?

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]: Yep, just a readability suggestion.

**Comment by Evangelink** [comment]
@mavasani I have 2 side questions:

1. Shall I open a ticket for the rename of the class? 

> I feel that with this user option the class is no longer strictly for collections and should be renamed as it is now recommending generic interface over interface for any functionality.

2. Shall I open a ticket for the doc update?

**Comment by mavasani** [comment]
Yes, both sound reasonable follow-up items to me. Thanks!
